### ISSUE TITLE
Introduce token-based actions

### ISSUE THREAD

**Issue by tiltec** [yunity/karrot-backend#275]
The following actions should require a token (valid for single use, limited time, received via e-mail) to be performed:

- Password reset (currently sends out a permanent password)
- Account delete
- Mail verification (already done, but called activation key)
- Mail change (already done, but called activation key)

We could add a new `Token` model with the fields `token`, `valid_before`, `is_used`. Then we can reference it in the user model with one `OneToOneField` (see Django docs for more details) per purpose.

This has some advantages over adding all fields to the user model:

- separation of concerns
- more DRY (don't repeat yourself)
- concise naming (otherwise we would have names like: `passwordreset_token_valid_before`)

I think this issue could be split in these steps:

1. Discuss and create the `Token` model
2. Refactor mail verification and mail change to use the token model
3. Implement account delete with token
4. Implement password reset with token

(Side note: while 1 & 2 should work without frontend changes and 3 is not implemented yet in the frontend, 4 would need some changes there as well. That's why I put it last)

**Comment by Josiah-DuBose** [comment]
Hello, I'd like to work on this issue if I could.

**Comment by tiltec** [comment]
Hi, sounds great! Be sure to join our Slack on slackin.yunity.org for further discussion.

**Comment by Josiah-DuBose** [comment]
is that slackin@yunity.org by chance?

**Comment by djahnie** [comment]
Just visit [slackin.yunity.org](https://slackin.yunity.org/) and send yourself an invitation to join our team chat, @Josiah-DuBose. On the yunity Slack you'll find a channel named #foodsaving-tool, where all the discussions regarding development of the tool take place. :)

**Comment by mr-kojo** [comment]
@tiltec suggested that we any term different from `Token` so that there is no confusion with [authentication tokens](http://www.django-rest-framework.org/api-guide/authentication/#tokenauthentication). I think that's a pretty reasonable idea so I made up my mind and came up with two candidates:

**_VerificationCode_**
PRO: Gives a pretty accurate idea of what it is meant for
CON: Some might think it solely represents an "e-mail verification code" although it can also stand for e.g. a "password reset verification code" (i.e., a code that verifies a requested password reset)

**_Key_**
PRO: Short
CON: Very general; _Key_ usually doesn't have the connotation that it can be used only once.

**Comment by mr-kojo** [comment]
Assuming we call the new model _VerificationCode_, I would suggest the following fields:

- `verification_code`: The token
- `timestamp_created`: The time of creation
- `type`: An enum field representing the type of the verification code (e.g., `mail_address_verification`, `password_reset`, `account_deletion`, etc.). We then create a mapping of API endpoints to types. Moreover, we create a configuration file which allows admins to set how long each type of verifcation code should be valid.
- `timestamp_used`: The time of use

**Comment by mr-kojo** [comment]
I also did a small survey on existing auth libraries. The following ones are under active development and look quite interesting:

- https://github.com/Tivix/django-rest-auth (authentication endpoints for Django REST: login/logout, mail verification, password reset, password change, user profile, Facebook/Twitter login)
- https://github.com/pennersr/django-allauth (very comprehensive, supporting OAuth2, developed with commercial support, for "plain" Django)
- https://github.com/Bouke/django-two-factor-auth (two-factor-authentication for Django: token generators, SMS, phone calls, and hardware tokens)

However, I have the impression that we better refrain from integrating an auth library into karrot at this stage. Most of what we currently need is already implemented so I guess we better spare ourselves the effort of integrating and configuring another library.

**Comment by tiltec** [comment]
`VerificationCode` sounds good to me, not as catchy as `Token`, but that's fine! I've seen it called [`hash`](https://github.com/getsentry/sentry/blob/fd056f95fe231e03af9b701cfda9b37d52a989eb/src/sentry/models/lostpasswordhash.py#L43) or [`key`](https://github.com/pennersr/django-allauth/blob/3fe584e1a3998e244e608d097ca4fececa77aa30/allauth/account/forms.py#L525) in other projects.
Maybe calling it `VerificationToken` or `VerificationKey` would also be an option? (We could do a quick vote on https://ukuvota.world/ among devs)

The suggested fields should be good, I especially like the idea of configurable validity.

For `type`, I would suggest to store it as string in the database, it's more meaningful than a-number-mapped-to-a-string. A bit more discussion on this is here: https://github.com/yunity/karrot-backend/issues/180#issuecomment-327246486

> I also did a small survey on existing auth libraries. 

Thanks! I think that's a separate topic, in fact there is already an issue:
https://github.com/yunity/karrot-frontend/issues/154

**Comment by mr-kojo** [comment]
> I think that's a separate topic, in fact there is already an issue

Oh, interesting, Facebook login has already been requested!

Well, I don't think it's a separate topic because if we decide to integrate some extensive auth library in the near future, it might be unnecessary to implement token-based actions now. For example, we might decide to use [django-rest-auth](https://github.com/Tivix/django-rest-auth) to solve #154 (as the library supports Facebook & Twitter login) as well as this very issue (as it supports mail verification & password reset) at the same time...

Of course, another option would be to implement token-based actions ourselves and at some time in the future look for a library that is just specialised on OAuth.

I'm not sure yet which option is better... I think I will have a closer look at django-rest-auth first.

**Comment by tiltec** [comment]
> Well, I don't think it's a separate topic because if we decide to integrate some extensive auth library in the near future, it might be unnecessary to implement token-based actions now.

Because the library might provide us with helpers to do password reset and email changes? Even if that is the case, I'm still unsure if it's good to integrate such an extensive library into our infrastructure. Might be quite much work, gives us less flexibility and might end up unmaintained. But it might also be the right choice...

> Of course, another option would be to implement token-based actions ourselves and at some time in the future look for a library that is just specialised on OAuth.

I'm tempted to go into this direction, but I don't want to discourage more research :)

**Comment by mr-kojo** [comment]
I had a closer look at the [django-rest-auth](https://github.com/Tivix/django-rest-auth) source code. It's a pretty light-weight library since it basically wraps existing Django functionality. Concretely, it instantiates Django forms and calls methods on them.
I think it's better if we don't use the library for now, but maybe we consider re-implementing some auth stuff in a similar fashion over time.

> VerificationCode sounds good to me, not as catchy as Token, but that's fine! I've seen it called hash or key in other projects. Maybe calling it VerificationToken or VerificationKey would also be an option?

I've resistance against _VerificationToken_ because a _Token_ might then be understood as a short form of both _AuthToken_ and _VerificationToken_.
And I prefer _VerificationCode_ over _VerificationKey_ because a _Key_ usually doesn't expire after being used.
So I'll just call it _VerificationCode_ 👍

**Comment by mr-kojo** [comment]
I would propose the following new API endpoints:

1. `POST: api/auth/verification_code/check` (replacing `api/auth/verify_mail`)
     - `type = (ACCOUNT_ACTIVATION | ACCOUNT_DELETION | EMAIL_CHANGE | PASSWORD_CHANGE | PASSWORD_RESET | INVITATION)`
     - `user` (required)
     - `code` (required)
2. `POST: api/auth/verification_code/resend_email` (replacing `api/auth/resend_verification`)
     - `type (ACCOUNT_ACTIVATION | ACCOUNT_DELETION | EMAIL_CHANGE | PASSWORD_CHANGE)`

Note that, in contrast to the current API, the user does not have to be logged in to validate (check) a verification code.

**Comment by tiltec** [comment]
> Note that, in contrast to the current API, the user does not have to be logged in to validate a verification code.

Great! I had that on my personal to-do list for some time already :)

What is the `user` field? User ID?

I think the `PASSWORD_CHANGE` type needs to accept the new password while validating, so the endpoint would become polymorphic (a bit cumbersome to document with Swagger).

**Comment by mr-kojo** [comment]
> What is the user field? User ID?

Yes, that would be the user's primary key.

> I think the PASSWORD_CHANGE type needs to accept the new password while validating, so the endpoint would become polymorphic

Why polymorphic? I imagine the following flow: 
1. `User.password` is set to the new password (via `change_password` endpoint)
2. `User.password_verified` is set to `False`
3. Verification code is sent via e-mail
4. On validation (via `verification_code/validate` endpoint), `User.password_verified` is set to `True`

So no need to pass the new password to the `verification_code/validate` endpoint, or am I wrong?

**Comment by tiltec** [comment]
Ah, that's a possibility! `PASSWORD_RESET` should work similar then?

**Comment by tiltec** [comment]
Why the double verification?

**Comment by mr-kojo** [comment]
Well, the second verification obviously has to happen.
The first one is just to enhance user experience: If we leave it away and the user opens a password reset link that has already expired, she/he will only be notified after submitting a new password.

**Comment by tiltec** [comment]
Similar to password change:

1. submit `password_reset` request as non-authed user, with body containing `e-mail` of the user to reset and `new_password` as the requested password
2. if `e-mail` exists, user gets e-mail with verification code. Password is hashed and saved as `User.password_to_reset`.
3. If user confirms with verification code, replace `User.password` with `User.password_to_reset`.
